src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1291 @@
+}
+
+/*
+ * Converts a simple ratbag macro to a set of at most two key codes and modifiers
+ *
+ * @return Number of key codes on success or a negative errno, -EPROTO if ratbag macro is too complex
+ */
+static int
+holtek8_keycodes_from_ratbag_macro(const struct ratbag_button_action *action, unsigned int *key1_out, unsigned int *key2_out, unsigned int *modifiers_out)
+{
+	const struct ratbag_macro *macro = action->macro;
+	const unsigned int num_keys = (unsigned int)ratbag_action_macro_num_keys(action);
+	unsigned int key1 = KEY_RESERVED;
+	unsigned int key2 = KEY_RESERVED;
+	unsigned int modifiers = 0;
+	unsigned int i;
+	unsigned int keys_pressed = 0;
+	unsigned int mods_pressed = 0;
+	unsigned int num_mods = 0;
+
+	if (!macro || action->type != RATBAG_BUTTON_ACTION_TYPE_MACRO)
+		return -EINVAL;
+
+	if (num_keys > 2)
+		return -EPROTO;
+
+	for (i = 0; i < MAX_MACRO_EVENTS; i++) {
+		const struct ratbag_macro_event *event = &macro->events[i];
+		if (event->type == RATBAG_MACRO_EVENT_NONE ||
+		    event->type == RATBAG_MACRO_EVENT_INVALID) {
+			break;
+		}
+		if (ratbag_key_is_modifier(event->event.key) && event->type == RATBAG_MACRO_EVENT_KEY_PRESSED) {
+			num_mods += 1;
+		}
+	}
+
+	for (i = 0; i < MAX_MACRO_EVENTS; i++) {
+		const struct ratbag_macro_event *event = &macro->events[i];
+
+		switch (event->type) {
+			case RATBAG_MACRO_EVENT_INVALID:
+				return -EINVAL;
+			case RATBAG_MACRO_EVENT_KEY_PRESSED:
+				switch (event->event.key) {
+				case KEY_LEFTCTRL: modifiers |= HOLTEK8_MODIFIER_LEFTCTRL; mods_pressed++; break;
+				case KEY_LEFTSHIFT: modifiers |= HOLTEK8_MODIFIER_LEFTSHIFT; mods_pressed++; break;
+				case KEY_LEFTALT: modifiers |= HOLTEK8_MODIFIER_LEFTALT; mods_pressed++; break;
+				case KEY_LEFTMETA: modifiers |= HOLTEK8_MODIFIER_LEFTMETA; mods_pressed++; break;
+				case KEY_RIGHTCTRL: modifiers |= HOLTEK8_MODIFIER_RIGHTCTRL; mods_pressed++; break;
+				case KEY_RIGHTSHIFT: modifiers |= HOLTEK8_MODIFIER_RIGHTSHIFT; mods_pressed++; break;
+				case KEY_RIGHTALT: modifiers |= HOLTEK8_MODIFIER_RIGHTALT; mods_pressed++; break;
+				case KEY_RIGHTMETA: modifiers |= HOLTEK8_MODIFIER_RIGHTMETA; mods_pressed++; break;
+				default:
+					if (key1 == KEY_RESERVED) {
+						key1 = event->event.key;
+					} else if (key2 == KEY_RESERVED) {
+						key2 = event->event.key;
+					} else {
+						return -EPROTO;
+					}
+					keys_pressed++;
+				}
+				break;
+			case RATBAG_MACRO_EVENT_NONE:
+			case RATBAG_MACRO_EVENT_KEY_RELEASED:
+				if (keys_pressed == num_keys && mods_pressed == num_mods) {
+					*modifiers_out = modifiers;
+					*key1_out = key1;
+					*key2_out = key2;
+					return (int)keys_pressed;
+				}
+				return -EPROTO;
+			case RATBAG_MACRO_EVENT_WAIT:
+				return -EPROTO;
+			default:
+				return -EINVAL;
+		}
+	}
+
+	return -EINVAL;
+}
+
+/*
+ * Converts raw macro events to ratbag macro and writes to a button
+ *
+ * @param raw A buffer of macro events at least HOLTEK8_MAX_MACRO_EVENTS of size
```

Comment:
I would prefer passing buffer size along with the pointer, I think it's a bit less error-prone.